src/bugpoint.rs

```diff
@@ -0,0 +1,312 @@
+
+        'inner_loop: for _ in 0..10000 {
+            progress.inc(1);
+            let mut func2 = func.clone();
+
+            let (msg, shrinked) = match phase.step(&mut func2) {
+                PhaseStepResult::Shrinked(msg) => (msg, true),
+                PhaseStepResult::Replaced(msg) => (msg, false),
+                PhaseStepResult::NoChange => continue 'inner_loop,
+                PhaseStepResult::NextPhase(msg, count) => {
+                    progress.set_message("done");
+                    progress.finish();
+                    progress =
+                        ProgressBar::with_draw_target(count as u64, ProgressDrawTarget::stdout());
+                    progress.set_style(
+                        ProgressStyle::default_bar()
+                            .template("{bar:60} {prefix:40} {pos:>4}/{len:>4} {msg}"),
+                    );
+                    progress.set_prefix(&format!("pass {} phase {}", pass_idx, msg));
+                    continue 'inner_loop;
+                }
+                PhaseStepResult::Finished => {
+                    progress.set_message("done");
+                    progress.finish();
+                    break 'inner_loop;
+                }
+            };
+
+            progress.set_message(&msg);
+
+            match check_for_crash(isa, &func2) {
+                Res::Succeed => {
+                    // Shrinking didn't hit the problem anymore, discard changes.
+                    //progress.println("succeeded");
+                    continue;
+                }
+                Res::Verifier(err) => {
+                    // Shrinking produced invalid clif, discard changes.
+                    //progress.println(format!("verifier error {}", err));
+                    continue;
+                }
+                Res::Panic => {
+                    // Panic remained while shrinking, make changes definitive.
+                    func = func2;
+                    if shrinked {
+                        was_reduced = true;
+                        if verbose {
+                            progress.println(format!("{}: shrink", msg));
+                        }
+                    } else {
+                        if verbose {
+                            progress.println(format!("{}: replace", msg));
+                        }
+                    }
+                }
+            }
+        }
+
+        if !was_reduced {
+            // No new shrinking opportunities have been found this pass. This means none will ever
+            // be found. Skip the rest of the passes over the function.
+            break 'outer_loop;
+        }
+    }
+
+    println!("\n{}", func);
+
+    println!(
+        "{} ebbs {} insts -> {} ebbs {} insts",
+        orig_ebb_count,
+        orig_inst_count,
+        ebb_count(&func),
+        inst_count(&func)
+    );
+}
+
+enum Res {
+    Succeed,
+    Verifier(String),
+    Panic,
+}
+
+fn check_for_crash(isa: &TargetIsa, func: &Function) -> Res {
```

Comment:
There can be several kinds of panics in the pipeline, and I'm afraid that checking for a crash as is could hide the initial one we're looking for, and instead cause a panic in an unrelated location. It would be nice to handle this case, e.g. by checking that the panic reason is the same as the one we observed during the first run I'm talking about above. Bonus points if, in this case, we can still have the clif causing the new panic.